### PR TITLE
Paper: Strassen

### PR DESCRIPTION
**Pull request by AbreitzmanSr**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

*Editor*: @mepa

*Reviewers*: @pushkarsathe89

### CODE REVIEW

**Comment by deniederhut** [comment]
@scoobies assign @mepa as editor

**Comment by scoobies** [comment]
Assigned! @mepa is now the editor

**Comment by mepa** [comment]
@scoobies build status

**Comment by scoobies** [comment]
{"AbreitzmanSr-strassen"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmpn2k6dclr'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: anthony_breitzman1\n*** Warning: /tmp/tmpn2k6dclr/publisher/../output/anthony_breitzman1/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/AbreitzmanSr-strassen.pdf", "build_status"=>"success", "build_timestamp"=>"25/05 00:17"}, "status"=>"success"}}

**Comment by mepa** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/ESPM254806.2021.00010 is OK
- 10.1016/j.jsc.2017.11.001 is OK
- 10.1007/BF02165411 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by deniederhut** [comment]
@scoobies add @pushkarsathe89 to reviewers

**Comment by scoobies** [comment]
@pushkarsathe89 added to the reviewers list!

**Comment by deniederhut** [comment]
@scoobies code of conduct

**Comment by scoobies** [comment]
Our CoC: https://www.scipy2023.scipy.org/code-of-conduct

**Comment by scoobies** [comment]
Reports of abusive or harassing behavior may be reported to scipy@enthought.com

File: papers/anthony_breitzman1/strassen.rst [hunk 1]
```diff
@@ -4,296 +4,413 @@
...
-If you wish to have a block quote, you can just indent the text, as in
-
-    When it is asked, What is the nature of all our reasonings concerning matter of fact? the proper answer seems to be, that they are founded on the relation of cause and effect. When again it is asked, What is the foundation of all our reasonings and conclusions concerning that relation? it may be replied in one word, experience. But if we still carry on our sifting humor, and ask, What is the foundation of all conclusions from experience? this implies a new question, which may be of more difficult solution and explication. :cite:`hume48`
-
-Dois in bibliographies
-++++++++++++++++++++++
-
-In order to include a doi in your bibliography, add the doi to your bibliography
-entry as a string. For example:
+A recent article [Arx21]_  suggests that Python is the *Lingua Franca* of machine learning and scientific computing because of powerful frameworks such as Numpy, SciPy, and TensorFlow. These libraries offer great flexibility while boosting the performance of Python because they are written in compiled C and C++.
```
Comment (pushkarsathe89) [inline papers/anthony_breitzman1/strassen.rst]: minor suggestion: article says its rapidly becoming, not it already is. modify wording

File: papers/anthony_breitzman1/strassen.rst [hunk 2]
```diff
@@ -4,296 +4,413 @@
...
 
-.. code-block:: bibtex
+In this short paper we present a modified Strassen-based [Stras1]_ algorithm for multiplying
```
Comment (pushkarsathe89) [inline papers/anthony_breitzman1/strassen.rst:27]: Can you add a short explanation as to why Strassen algorithm is of interest, and an example that motivates the need for a faster algorithm for integers? That would help motivate the work for readers.

File: papers/anthony_breitzman1/strassen.rst [hunk 3]
```diff
@@ -4,296 +4,413 @@
...
 
-   g(x) &=& \int_0^\infty f(x) dx \\
-        &=& \ldots
+   \begin{table*}[htbp]
+   \centering
+   \begin{tabular}{|l|r|r|r|r|r|r|}
+   \toprule
+       & \multicolumn{2}{c|}{Numpy} & \multicolumn{2}{c|}{Strassen 1} & \multicolumn{2}{c|}{Standard Multiply} \\
+   \midrule
+   \multicolumn{1}{|p{3.645em}|}{Matrix Size} & \multicolumn{1}{p{4.43em}|}{Time \ (seconds) } & \multicolumn{1}{p{4em}|}{Current/ Previous} & \multicolumn{1}{p{4.43em}|}{Time \  (seconds) } & \multicolumn{1}{p{4em}|}{Current/ Previous} & \multicolumn{1}{p{4.43em}|}{Time \ (seconds) } & \multicolumn{1}{p{4em}|}{Current/ Previous} \\
```
Comment (pushkarsathe89) [inline papers/anthony_breitzman1/strassen.rst:101]: If you want, you can use images of tables instead. Either way is good.

File: papers/anthony_breitzman1/strassen.rst [hunk 4]
```diff
@@ -4,296 +4,413 @@
...
+   Timing of Multiple Algorithms Relative to Strassen128 on Windows 11 with Core i7 @ 3.0 GHz.
+   :label:`figUpstairs5`
 
-   \end{table*}
+   
+Conclusions
+-----------
 
-Perhaps we want to end off with a quote by Lao Tse [#]_:
+Numpy is a Python library which is widely used in the math and scientific community because of its speed. In this paper we presented a Strassen type algorithm for multiplying large matrices with integer entries.  For integer matrices with row dimension or column dimension in the thousands the algorithm can be 8 to 30 times faster than Numpy. The algorithm is the standard Strassen divide and conquer algorithm but it crosses over to Numpy when either the row or column dimension of one of the matrices drops below 128.  The algorithm was tested on a MacBook, an I7 based Windows machine as well as a Linux machine running a Xeon processor with similar results.  Although there is no apparent advantage for matrices with real entries, there are a number of applications for matrices with integer coefficients.
```
Comment (pushkarsathe89) [inline papers/anthony_breitzman1/strassen.rst]: Is there a known reason why there is no benefit for matrices with real entries?

File: papers/anthony_breitzman1/strassen.rst [hunk 5]
```diff
@@ -4,296 +4,413 @@
...
+      return(np.vstack((m,x)))
+    
+   def padColumn(m):
+   """add column of zeros to right of matrix"""
+      x = []
+      for i in range(len(m)):
+        x.append(0)
+      return(np.hstack((m,np.vstack(x))))
+
+Since the padded rows (or columns) will need to be removed from the product at each level one might wonder whether padding once to a power of 2 would be more efficient?  For example, a matrix with 17 rows and 17 columns will be padded to :math:`18 \times 18`, but then each of its :math:`9 \times 9` submatrices will be padded to :math:`10 \times 10` which will require :math:`5 \times 5` submatrices to be padded and so on.  Cases like this could be avoided by padding the original matrix to :math:`32 \times 32`.  This was tested however, and it was found that padding of a single row at multiple levels of recursion is considerably faster than padding to the next power of 2.
```
Comment (pushkarsathe89) [inline papers/anthony_breitzman1/strassen.rst:255]: How about using powers of 3? or 5, etc...Perhaps an interesting question to explore in the future?

File: papers/anthony_breitzman1/strassen.rst (hunk 1 above)
Comment (AbreitzmanSr) [reply papers/anthony_breitzman1/strassen.rst]: will do.  thanks.

File: papers/anthony_breitzman1/strassen.rst (hunk 2 above)
Comment (AbreitzmanSr) [reply papers/anthony_breitzman1/strassen.rst:27]: yes.  good idea.

File: papers/anthony_breitzman1/strassen.rst (hunk 3 above)
Comment (AbreitzmanSr) [reply papers/anthony_breitzman1/strassen.rst:101]: I'm a native LaTex writer not RST.  When I tried an image it wanted to make it a figure and not a table.  If you have a tip I can give it a try.

File: papers/anthony_breitzman1/strassen.rst (hunk 4 above)
Comment (AbreitzmanSr) [reply papers/anthony_breitzman1/strassen.rst]: I tried it with reals and saw no significant benefit.

I was surprised when I changed the matrix coefficients to integers and saw a huge difference.  (e.g. Strassen 30 times faster than regular numpy for very large matrices).

File: papers/anthony_breitzman1/strassen.rst (hunk 5 above)
Comment (AbreitzmanSr) [reply papers/anthony_breitzman1/strassen.rst:255]: Strassen requires even dimension matrices for the divide and conquer.  In the example above of 17 x 17 the nearest power of 2 is 32 so we would have 32x32 then 16x16 then 8x8 etc. during the divide and conquer.  

Power of 3 or 5 wouldn't work.  For example the nearest power of 3 to 17 is 27 so we would still need to pad it to 28 to get an even dimension 28 x 28 then 14 x 14 then 7 x 7 which would need to be padded again, so we're no better off then our original 17 x 17 matrix which needs multiple padding steps.

**Comment by mepa** [comment]
@scoobies mark pending comment

**Comment by AbreitzmanSr** [comment]
I've updated the draft based on the reviewers comments (which were helpful) but how do I resubmit the draft?  Do I do another pull request on the current branch or does it go somewhere else.

(There's a good chance this is a stupid question because I am a github novice.)

Thanks!

**Comment by deniederhut** [comment]
> Do I do another pull request

Nope! Push those commits to your fork, and those changes will show up here on GH. You'll need to do `@scoobies build paper` or click the refresh button on procbuild to recompile the PDF on the build server.

**Comment by AbreitzmanSr** [comment]
ok, that seems to have worked.  I see the new version on your build server.
Thanks!
________________________________
From: Dillon Niederhut ***@***.***>
Sent: Friday, July 14, 2023 4:33 PM
To: scipy-conference/scipy_proceedings ***@***.***>
Cc: Breitzman, Anthony F ***@***.***>; Author ***@***.***>
Subject: [EXTERNAL] Re: [scipy-conference/scipy_proceedings] Paper: Strassen (PR #802)


Do I do another pull request

Nope! Push those commits to your fork, and those changes will show up here on GH. You'll need to do @scoobies build paper or click the refresh button on procbuild to recompile the PDF on the build server.

**Comment by mepa** [comment]
Hi @pushkarsathe89, how are you feeling about the updates from the author?

**Comment by mepa** [comment]
@scoobies check references

**Comment by scoobies** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1109/ESPM254806.2021.00010 is OK
- 10.1016/j.jsc.2017.11.001 is OK
- 10.1007/BF02165411 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mepa** [comment]
Hey @pushkarsathe89! The [deadline](https://github.com/scipy-conference/scipy_proceedings#reviewer-deadlines) for final recommendations was yesterday. Do you have a final recommendation to make? Is this paper ready for inclusion in the conference proceedings?

**Comment by pushkarsathe89** [comment]
Hi @mepa , apologies for not clearing it up earlier. Its ready. Most of my questions were answered and my comments were mostly optional changes which don't affect the conclusion.

**Comment by mepa** [comment]
@scoobies mark ready

**Review by deniederhut — APPROVED** [review]
Thanks @pushkarsathe89 for your helping us get this paper ready for publication! And thanks @AbreitzmanSr for... (checks notes)... submitting two papers this year!

tenor-1742029395

**Pull request merged**